bonus/hkd.md

```diff
@@ -0,0 +1,367 @@
+# HKD - Higher-Kinded Data
+
+This file is literate haskell. If you have
+[`markdown-unlit`](https://hackage.haskell.org/package/markdown-unlit)
+installed, you can extract the code by running the following command:
+
+```
+$ markdown-unlit -h mtl.md mtl.md mtl.hs
+```
+
+This will allow you to work through the exercises using GHCi.
+
+In this document, we will explore a pattern called "higher-kinded
+data", which is when a data structure is parameterised by a
+`Functor`. This is becoming increasingly common in libraries like
+[`waargonaut`](https://hackage.haskell.org/package/waargonaut) (which
```

Comment:
I've not actually tried that out, but leveraging the floating `f` would feel very validating considering how much of PITA it was sometimes....

Reply:
I hope so. I've never needed it but it's a lib the students will be familiar with.